### PR TITLE
Yair/yagever

### PR DESCRIPTION
**Pull request by OrShtrauch**

### CODE REVIEW

File: sniffer.c [hunk 1]
```diff
@@ -0,0 +1,312 @@
...
+
+// declaring those variables globally for use in cleanup method
+static nflog_handle_t *handle;
+static nflog_g_handle_t *group_handle;
+static FILE *log_fd;
+
+int iptables(const char *iptables_path, int delete, char *nflog_group, char *dport) {
+    char *argv[] = {
+        iptables_path,
+        (delete == 1) ? "-D" : "-A",
```
Comment (yairtabib) [inline sniffer.c]: delete ? "-D" : "-A"

File: sniffer.c [hunk 2]
```diff
@@ -0,0 +1,312 @@
...
+
+    printf("running iptables command: ");
+    for (int i = 0; argv[i]!= NULL; i++) {
+        printf("%s ", argv[i]);
+    }
+    printf("\n");
+
+    pid_t pid = fork();
+    if (pid < 0) {
+        printf("failed running fork");
```
Comment (yairtabib) [inline sniffer.c]: perror

File: sniffer.c [hunk 3]
```diff
@@ -0,0 +1,312 @@
...
+#include <libnetfilter_log/libnetfilter_log.h>
+#include <netinet/ip.h>
+#include <netinet/udp.h>
+#include <signal.h>
+#include <stdarg.h>
+#include <stdio.h>
+#include <stdlib.h>
+#include <string.h>
+#include <unistd.h>
+
```
Comment (yairtabib) [inline sniffer.c]: create define of errors and success

File: sniffer.c [hunk 4]
```diff
@@ -0,0 +1,312 @@
...
+        return 1;
+    } else if (pid == 0) {
+        execv(iptables_path, argv);
+        perror("Exec failed\n");
+        exit(1);
+    } else {
+        wait(NULL);
+    }
+
+    return 0;
```
Comment (yairtabib) [inline sniffer.c]: pid_t pid = fork();
    if (pid < 0) {
        printf("failed running fork");
        return 1;
    } 

    if (pid == 0) {
        execv(iptables_path, argv);
        perror("Exec failed\n");
        exit(1);
    } else {
        wait(NULL);
    }

File: sniffer.c [hunk 5]
```diff
@@ -0,0 +1,312 @@
...
+
+        dns_payload = (char *)(payload + IPV6_DNS_PAYLOAD_OFFSET);
+        dns_payload_len = payload_len - IPV6_DNS_PAYLOAD_OFFSET;
+    }
+
+    parse_domain(dns_payload, dns_payload_len, response, &seek);
+    parse_query_type(dns_payload, seek, response);
+}
+
+static int callback(nflog_g_handle_t *group_handle, nfgenmsg_t *nfmsg, nflog_data_t *nfa, void *data) {
```
Comment (yairtabib) [inline sniffer.c]: cb_&_wtf_it_does

File: sniffer.c [hunk 6]
```diff
@@ -0,0 +1,333 @@
...
+        perror("failed running fork");
+        return EXIT_FAILURE;
+    }
+    
+    if (pid == 0) {
+        execv(iptables_path, argv);
+        perror("Exec failed\n");
+        exit(1);
+    } else {
+        wait(NULL);
```
Comment (orgoldshmidt) [inline sniffer.c]: receive return value from the forked process.

File: sniffer.c [hunk 7]
```diff
@@ -0,0 +1,333 @@
...
+            break;
+
+        /*
+            An example for the domain section in the dns payload
+            03 'w' 'w' 'w' 07 'e' 'x' 'a' 'm' 'p' 'l' 'e' 03 'c' 'o' 'm' 00
+            each section is prefix with each length, followed by a '.'
+            character so, ill fetch the length of the current section, then
+            loop for this length and print each char
+        */
+        for (int j = 1; j <= section_length; j++) {
```
Comment (orgoldshmidt) [inline sniffer.c]: Invalid in C!!! You can't declare in a loop.

File: dns_sniffer/include/dns_sniffer.h [hunk 8]
```diff
@@ -0,0 +1,86 @@
...
+#define BUFFER_SIZE 4096
+#define DNS_PORT 53
+
+#define IP_VERSION_MASK 0x0f
+#define MAX_PKT_SIZE 0xffff
+
+#define IPV4_DNS_PAYLOAD_OFFSET 28
+#define IPV6_DNS_PAYLOAD_OFFSET 48
+
+#define LOG_FILE_PATH "./log"
```
Comment (IdoAd) [inline dns_sniffer/include/dns_sniffer.h]: Full path in code

File: dns_sniffer/include/dns_sniffer.h [hunk 9]
```diff
@@ -0,0 +1,86 @@
...
+#define MAX_LINE_SIZE 1024
+#define DOMAIN_MAX_SIZE 256
+#define IP_MAX_SIZE 40  // using ipv6 size
+
+typedef struct nflog_handle nflog_handle_t;
+typedef struct nflog_g_handle nflog_g_handle_t;
+typedef struct nfgenmsg nfgenmsg_t;
+typedef struct nflog_data nflog_data_t;
+
+typedef enum {
```
Comment (IdoAd) [inline dns_sniffer/include/dns_sniffer.h]: Globally defined very general naming, add prefix such as DNS_OTHER or move to .c

File: dns_sniffer/include/dns_sniffer.h [hunk 10]
```diff
@@ -0,0 +1,86 @@
...
+    A = 1,
+    AAAA = 28,
+    CNAME = 5,
+    OTHER = -1
+} qtype_t;
+
+typedef enum {
+    IPV4 = 4,
+    IPV6 = 6
+} ip_version_t;
```
Comment (IdoAd) [inline dns_sniffer/include/dns_sniffer.h]: At SamDaemon its not our syntax for some reason we always write struct X \ enum X

File: dns_sniffer/include/dns_sniffer.h [hunk 11]
```diff
@@ -0,0 +1,86 @@
...
+dns_sniffer_exit_status_t init_dns_sniffer(dns_sniffer_t *sniffer, uint16_t nflog_group);
+
+void parse_domain(const char *dns_payload, int dns_payload_len, dns_response_t *out, int *seek);
+
+void parse_query_type(char *dns_payload, int question_start, dns_response_t *out);
+
+void parse_dns_packet(char *payload, int payload_len, dns_response_t *out);
+
+int cb_handle_dns_packet(nflog_g_handle_t *group_handle, nfgenmsg_t *nfmsg, nflog_data_t *nfa, void *data);
+
```
Comment (IdoAd) [inline dns_sniffer/include/dns_sniffer.h:80]: All functions definitions above should be static

File: dns_sniffer/include/dns_sniffer.h (hunk 9 above)
Comment (IdoAd) [reply dns_sniffer/include/dns_sniffer.h]: I believe that generally all definitions that are not needed publicly should be in the .c file
from one side when included if not used the wont be compiled there from the other it is cleaner "name-spacing" only what is needed for the API usage is there..
it is possible to create .h for public usage and .h for private..

File: dns_sniffer/include/dns_sniffer.h [hunk 12]
```diff
@@ -0,0 +1,86 @@
...
+typedef enum {
+    DS_NFLOG_OPEN_ERROR_EXIT_CODE,
+    DS_NFLOG_UNBINDING_PF_ERROR_EXIT_CODE,
+    DS_NFLOG_BINDING_PF_ERROR_EXIT_CODE,
+    DS_NFLOG_BIND_GROUP_ERROR_EXIT_CODE,
+    DS_NFLOG_SET_MODE_ERROR_EXIT_CODE,
+    DS_NFLOG_REGISTER_CALLBACK_ERROR_EXIT_CODE,
+    DS_NFLOG_OPEN_FD_ERROR_EXIT_CODE,
+    DS_GENERAL_FAILURE_EXIT_CODE,
+    DS_SUCCESS_EXIT_CODE = 0
```
Comment (IdoAd) [inline dns_sniffer/include/dns_sniffer.h:48]: Did you mean that the previous parameters are negative? they are not DS_NFLOG_OPEN_ERROR_EXIT_CODE = 0, DS_NFLOG_UNBINDING_PF_ERROR_EXIT_CODE =1...... and DS_SUCCESS_EXIT_CODE = 0 as you wished...
so you can't differ between DS_NFLOG_OPEN_ERROR_EXIT_CODE DS_SUCCESS_EXIT_CODE

File: dns_sniffer/include/dns_sniffer.h [hunk 13]
```diff
@@ -0,0 +1,86 @@
...
+
+int cb_handle_dns_packet(nflog_g_handle_t *group_handle, nfgenmsg_t *nfmsg, nflog_data_t *nfa, void *data);
+
+dns_sniffer_exit_status_t register_packet_handler(dns_sniffer_t *sniffer, dns_callback_data_t *callback_data);
+
+dns_sniffer_exit_status_t start_dns_sniffer(dns_sniffer_t *sniffer);
+
+void close_dns_sniffer(dns_sniffer_t *sniffer);
+
+#endif
```
Comment (IdoAd) [inline dns_sniffer/include/dns_sniffer.h:87]: We do leave empty line at the end of each file

File: dns_sniffer/src/dns_sniffer.c [hunk 14]
```diff
@@ -0,0 +1,182 @@
+#include "dns_sniffer.h"
+
+#include <arpa/inet.h>
+#include <libnetfilter_log/libnetfilter_log.h>
+#include <stdlib.h>
+#include <string.h>
+#include <unistd.h>
```
Comment (IdoAd) [inline dns_sniffer/src/dns_sniffer.c:7]: I checked it following this, best practice for #include order in .c file is
Recommended Order:

Header File Corresponding to the .c File:  Start by including the header file that directly matches the current C source file (e.g., foo.c includes foo.h). This ensures that the header file is self-contained and doesn't rely on external headers being included first.

Local Project Headers: Include other header files from your project, grouped by module or component. This helps keep dependencies within your project clear.

External Library Headers: Include headers from external libraries (e.g., standard C library, third-party libraries) in a separate group.

System Headers: Include standard system headers (e.g., <stdio.h>, <stdlib.h>) last.

Example:
#include "foo.h"    // Corresponding header
#include "utils.h"  // Local header
#include "network.h" // Local header

#include <pthread.h>  // External library
#include <math.h>  // External library

#include <stdio.h>   // System header
#include <string.h>  // System header

File: dns_sniffer/include/dns_sniffer.h (hunk 11 above)
Comment (IdoAd) [reply dns_sniffer/include/dns_sniffer.h:80]: and removed from .h as private functions

File: firewall/include/firewall.h [hunk 15]
```diff
@@ -0,0 +1,27 @@
+#ifndef __FIREWALL_H__
+#define __FIREWALL_H__
+
+#include <stdint.h>
+
+#define DNS_PORT_STR "53"
```
Comment (IdoAd) [inline firewall/include/firewall.h:6]: as in previous header file.
My convention is if it is not needed to be published as other files use it too, it should be in the .c file.
private function should be static and not in .h

it makes it readability for the client of this code, only needed definitions and functions are visible in .h

File: main.c [hunk 16]
```diff
@@ -0,0 +1,100 @@
...
+
+    printf("opening log file\n");
+    callback_data.output_fd = fopen(LOG_FILE_PATH, LOG_FILE_MODE);
+    if (!callback_data.output_fd) {
+        printf("Error opening log file\n");
+        return FAILURE_EXIT;
+    }
+
+    printf("Setting up iptables rules\n");
+    setup_iptables_rules();
```
Comment (IdoAd) [inline main.c:90]: main.c is the client of firewall and dns_sniffer
 he doesn't need to do everything
 why init_dns_sniffer that you give him NFLOG_GROUP cant do setup_iptables_rules() probably register_packet_handler to..
 Generally if the client not really need the control make it easier for him to use it..

File: main.c [hunk 17]
```diff
@@ -0,0 +1,100 @@
...
+
+    printf("Initializing DNS sniffer\n");
+    exit_code = init_dns_sniffer(&sniffer, NFLOG_GROUP);
+    if (exit_code != DS_SUCCESS_EXIT_CODE) {
+        printf("Failed to initialize sniffer, exit code: %d\n", exit_code);
+        cleanup();
```
Comment (IdoAd) [inline main.c]: int rc = OK;
....
rc = FAILURE_EXIT;
goto cleanup:
....
cleanup:
return rc;

Comment (IdoAd) [reply main.c]: always try one point of failure, one point of cleanup..

File: main.c [hunk 18]
```diff
@@ -0,0 +1,100 @@
...
+
+#define IPV4 4
+#define IPV6 6
+
+#define NFLOG_GROUP 3
+
+#define SUCCESS_EXIT 0
+#define FAILURE_EXIT 1
+
+dns_sniffer_t sniffer;
```
Comment (IdoAd) [inline main.c]: if really needed to be hold by dns_sniffer client, should be static and should be renamed g_sniffer or such.
note for dns_sniffer functions with the variable names sniffer

File: dns_sniffer/include/dns_sniffer.h (hunk 11 above)
Comment (IdoAd) [reply dns_sniffer/include/dns_sniffer.h:80]: Naming convention, public functions should have feature prefix
dns_sniffer_register_packet_handler...
dns_sniffer_start
dns_sniffer_close..

private static functions doesn't need the prefix

File: main.c [hunk 19]
```diff
@@ -0,0 +1,100 @@
...
+           response->domain,
+           (response->ip_version == IPV4) ? "IPv4" : "IPv6",
+           (response->query_type == A) ? "A" : (response->query_type == AAAA) ? "AAAA"
+                                           : (response->query_type == CNAME)  ? "CNAME"
+                                                                              : "Unknown");
+}
+
+void signal_handler(int signum) {
+    printf("got signal %d\n", signum);
+    cleanup();
```
Comment (IdoAd) [inline main.c]: you are in signal handler, do minimal work, raise flag and let the program do the cleanup

File: firewall/include/firewall.h (hunk 15 above)
Comment (OrShtrauch) [reply firewall/include/firewall.h:6]: unit16_t is used in  public func sig

**Pull request merged**